### ISSUE TITLE
Respond to chunk requests once (per some time interval)

### ISSUE THREAD

**Issue by birchmd** [near/nearcore#3830]
Nodes will request chunk parts they are missing when they get a block which includes a chunk they do not yet have (either have their part for or have reconstructed if it is for a shard they track). These requests are re-sent relatively frequently to ensure they are delivered and responded to. However, presently nodes will always respond to these requests, even if they have responded in the past. This can lead to redundant messages being sent over the network and could potentially contribute to congestion which is slowing down the network.

In this issue we want to add logic to node such they they will only respond to a request once per some time period (which can then be tuned for performance).

**Comment by bowenwang1996** [comment]
I think it may be better to do this on the network level, for all requests that require responses.

**Comment by birchmd** [comment]
Update on this issue:

We tried not accepting duplicate messages at the network level, remembering which we had seen before using the sha256 hash of the message ([commit](https://github.com/birchmd/nearcore/commit/21f2c55248954a150142b11d1513e9eeadb631ae)). But the hashing of messages introduced more overhead than ignoring duplicates saved.

We tried storing the raw message bytes as well ([commit](https://github.com/birchmd/nearcore/commit/53efc79ae3e25a77bc1d9c50227b51916ff66f1d)). This seemed to improve performance a little bit, though made the memory use much higher, probably too high to be an acceptable solution for production.

We tried using `xxHash` instead of sha256 since it is a faster hash function, but this also didn't seem to help performance very much. However, this might still be worth moving to production as a simple DOS prevention measure.

**Comment by birchmd** [comment]
Update: upgrading to the latest actix/tokio versions (see https://github.com/near/nearcore/issues/3868 and https://github.com/near/nearcore/pull/3869) seems to have a much greater impact on performance (may even allow sustaining 4k tps on 8 shards). The possibility of DOS attacks via flooding nodes with messages (especially since the p2p network is fully open) still exists, but that can be addressed as a general stability/security fix, rather than a sharding performance improvement. We also have an issue which is likely to reduce the amount of network traffic generated by (honest) non-validating nodes (see https://github.com/near/nearcore/issues/3890).

Given the above, I am closing this issue.

**Comment by bowenwang1996** [comment]
I don't think we should close this issue given that it is not done and we still need to do it.
